cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go

```diff
@@ -70,34 +83,80 @@ func (o *checkCapacityProvClass) Provision(
 	daemonSets []*appsv1.DaemonSet,
 	nodeInfos map[string]*schedulerframework.NodeInfo,
 ) (*status.ScaleUpStatus, errors.AutoscalerError) {
-	if len(unschedulablePods) == 0 {
-		return &status.ScaleUpStatus{Result: status.ScaleUpNotTried}, nil
-	}
+	combinedStatus := NewCombinedStatusSet()
+	provisioningRequestsProcessed := 0
+	startTime := time.Now()
 
-	prs := provreqclient.ProvisioningRequestsForPods(o.client, unschedulablePods)
-	prs = provreqclient.FilterOutProvisioningClass(prs, v1.ProvisioningClassCheckCapacity)
-	if len(prs) == 0 {
-		return &status.ScaleUpStatus{Result: status.ScaleUpNotTried}, nil
-	}
-	// Pick 1 ProvisioningRequest.
-	pr := prs[0]
 	o.context.ClusterSnapshot.Fork()
 	defer o.context.ClusterSnapshot.Revert()
 
-	scaleUpIsSuccessful, err := o.checkcapacity(unschedulablePods, pr)
-	if err != nil {
-		return status.UpdateScaleUpError(&status.ScaleUpStatus{}, errors.NewAutoscalerError(errors.InternalError, "error during ScaleUp: %s", err.Error()))
-	}
-	if scaleUpIsSuccessful {
-		return &status.ScaleUpStatus{Result: status.ScaleUpSuccessful}, nil
+	for len(unschedulablePods) > 0 {
+		prs := provreqclient.ProvisioningRequestsForPods(o.client, unschedulablePods)
+		prs = provreqclient.FilterOutProvisioningClass(prs, v1.ProvisioningClassCheckCapacity)
+		if len(prs) == 0 {
+			break
+		}
+
+		// Pick 1 ProvisioningRequest.
+		pr := prs[0]
+
+		scaleUpIsSuccessful, err := o.checkcapacity(unschedulablePods, pr)
```

Comment:
IIUC, we want to move on to evaluating the next PR whether the scheduling is successful or not. `checkcapacity` calls `TrySchedulePods`, which will keep scheduling the provided pods until the first failure. Whether all pods schedule or not, all scheduling results prior to the first failure are persisted in the cluster snapshot.

If all pods schedule then this seems to be what we want. But if one of them does fail and we mark the PR as Provisioned=False,  shouldn't we roll back the scheduling of previous pods from this PR in the cluster snapshot? I.e. shouldn't `checkcapacity` fork the snapshot and commit/revert based on the scheduling result?

Reply:
Fixed. I accidentally removed it at some point.

Reply:
Why not fork and commit/revert inside `checkcapacity`? IMO it'd both make this callsite more readable, and `checkcapacity` would have clearer semantics -  "persists the scheduling whenever all pods can be scheduled, and doesn't change the snapshot otherwise".

The current pattern is quite error prone because you have to remember to commit/revert in every relevant codepath at the callsite. E.g. don't we forget to do either if `o.checkCapacityBatchProcessing` is False and `o.checkcapacity()` returns an error?

Reply:
Fixed.